src/main/scala/cognite/spark/v1/FlexibleDataModelRelationUtils.scala

```diff
@@ -730,185 +750,139 @@ object FlexibleDataModelRelationUtils {
-                                    |""".stripMargin))
-    }
-  }
+      val asBigDecimal = BigDecimal(String.valueOf(n))
+      safeConvert(asBigDecimal)
+    }.toEither
+      .leftMap(exception =>
+        new CdfSparkException(
+          s"""Error parsing value for field '$propertyName'.
+                                 |Expecting a $expectedType but found '${String.valueOf(n)}'
+                                 |""".stripMargin,
+          exception
+      ))
 
-  private def tryAsFloatSeq(ns: Seq[Any], propertyName: String): Either[CdfSparkException, Seq[Float]] =
+  private def tryConvertNumberSeq[T](
+      ns: Seq[Any],
+      propertyName: String,
+      expectedType: String,
+      safeConvert: (BigDecimal) => T): Either[CdfSparkException, Seq[T]] =
     Try {
       skipNulls(ns).map { n =>
-        val bd = BigDecimal(String.valueOf(n))
-        if (bd.isDecimalFloat) {
-          bd.floatValue
-        } else {
-          throw new IllegalArgumentException(s"'${String.valueOf(n)}' is not a valid Float")
-        }
+        val asBigDecimal = BigDecimal(String.valueOf(n))
+        safeConvert(asBigDecimal)
       }
-    } match {
-      case Success(value) => Right(value)
-      case Failure(e) =>
+    }.toEither
+      .leftMap(exception => {
         val seqAsStr = ns.map(String.valueOf).mkString(",")
-        Left(new CdfSparkException(s"""Error parsing value for field '$propertyName'.
-                                      |Expecting an Array[Float] but found '[$seqAsStr]'
-                                      |${e.getMessage}
-                                      |""".stripMargin))
-    }
+        new CdfSparkException(
+          s"""Error parsing value for field '$propertyName'.
+                                 |Expecting a $expectedType but found '$seqAsStr'
+                                 |""".stripMargin,
+          exception
+        )
+      })
 
-  private def tryAsDouble(n: Any, propertyName: String): Either[CdfSparkException, Double] = {
-    val nAsStr = String.valueOf(n)
-    val bd = BigDecimal(nAsStr)
-    if (bd.isDecimalDouble) {
-      Right(bd.doubleValue)
-    } else {
-      Left(new CdfSparkException(s"""Error parsing value for field '$propertyName'.
-                                    |Expecting a Double but found '$nAsStr'
-                                    |""".stripMargin))
-    }
-  }
+  private def tryAsDirectNodeRelationList(
+      value: Seq[Row],
+      propertyName: String,
+      instanceSpace: Option[String]): Either[CdfSparkException, Vector[DirectRelationReference]] =
+    skipNulls(value).toVector
+      .traverse(extractDirectRelations(propertyName, "Direct Node Relation", instanceSpace, _))
 
-  private def tryAsDoubleSeq(
-      ns: Seq[Any],
-      propertyName: String): Either[CdfSparkException, Seq[Double]] =
+  private def extractDirectRelations(
+      propertyName: String,
+      descriptiveName: String,
+      defaultSpace: Option[String],
```

Comment:
```suggestion
      instanceSpace: Option[String],
```
at least the calling method uses that name